internal/state/mailbox_search.go

```diff
@@ -4,425 +4,581 @@ import (
 	"net/mail"
+	"runtime"
 	"strings"
+	"sync/atomic"
 	"time"
 
 	"github.com/ProtonMail/gluon/imap"
 	"github.com/ProtonMail/gluon/internal/contexts"
-	"github.com/ProtonMail/gluon/internal/db"
 	"github.com/ProtonMail/gluon/internal/db/ent"
 	"github.com/ProtonMail/gluon/internal/parser/proto"
 	"github.com/ProtonMail/gluon/rfc822"
-	"github.com/bradenaw/juniper/xslices"
 	"golang.org/x/text/encoding"
 )
 
+var totalActiveSearchRequests int32
+
 func (m *Mailbox) Search(ctx context.Context, keys []*proto.SearchKey, decoder *encoding.Decoder) ([]uint32, error) {
-	snapMessages := m.snap.getAllMessages()
+	var mapFn func(snapMsgWithSeq) uint32
 
-	messages, err := doSearch(ctx, m, snapMessages, keys, decoder)
+	if contexts.IsUID(ctx) {
+		mapFn = func(s snapMsgWithSeq) uint32 {
+			return uint32(s.UID)
+		}
+	} else {
+		mapFn = func(s snapMsgWithSeq) uint32 {
+			return uint32(s.Seq)
+		}
+	}
+
+	op, err := buildSearchOpListWithKeys(m, keys, decoder)
 	if err != nil {
 		return nil, err
 	}
 
-	return xslices.Map(messages, func(msg snapMsgWithSeq) uint32 {
-		if contexts.IsUID(ctx) {
-			return uint32(msg.UID)
+	msgCount := m.snap.len()
+
+	result := make([]uint32, msgCount)
+
+	activeSearchRequests := atomic.AddInt32(&totalActiveSearchRequests, 1)
+	defer atomic.AddInt32(&totalActiveSearchRequests, -1)
+
+	parallelism := runtime.NumCPU() / int(activeSearchRequests)
+
+	if err := parallel.DoContext(ctx, parallelism, msgCount, func(ctx context.Context, i int) error {
+		msg, ok := m.snap.messages.getWithSeqID(imap.SeqID(i + 1))
+		if !ok {
+			return nil
+		}
+
+		matches, err := applySearch(ctx, m, msg, op)
+		if err != nil {
+			return err
+		}
+
+		if matches {
+			result[i] = mapFn(msg)
 		}
 
-		return uint32(msg.Seq)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return xslices.Filter(result, func(v uint32) bool {
+		return v != 0
 	}), nil
 }
 
-func doSearch(ctx context.Context, m *Mailbox, candidates []snapMsgWithSeq, keys []*proto.SearchKey, decoder *encoding.Decoder) ([]snapMsgWithSeq, error) {
-	for _, key := range keys {
-		filtered, err := m.matchSearchKey(ctx, candidates, key, decoder)
+func buildSearchData(ctx context.Context, m *Mailbox, op *buildSearchOpResult, message snapMsgWithSeq) (searchData, error) {
+	data := searchData{message: message}
+
+	if op.needsMessage {
+		dbm, err := db.ReadResult(ctx, m.state.db(), func(ctx context.Context, client *ent.Client) (*ent.Message, error) {
+			return db.GetMessageDateAndSize(ctx, client, message.ID.InternalID)
+		})
+		if err != nil {
+			return searchData{}, err
+		}
+
+		data.dbMessage.size = dbm.Size
+		data.dbMessage.date = dbm.Date
+	}
+
+	if op.needsLiteral {
+		l, err := m.state.getLiteral(message.ID.InternalID)
 		if err != nil {
-			return nil, err
+			return searchData{}, err
 		}
 
-		candidates = filtered
+		data.literal = l
 	}
 
-	return candidates, nil
+	if op.needsHeader {
+		headerBytes, _ := rfc822.Split(data.literal)
```

Comment:
could probably optimize this with a separate `rfc822.SplitHeader`  and `rfc822.SplitBody` that doesn't unnecessarily return the header/body?

Reply:
It's returning sub-slices, no copies are performed.